src/test/data/script_tests.json

```diff
@@ -2527,6 +2527,344 @@
+    "P2SH,WITNESS,TAPROOT",
+    "OK",
+    "TAPSCRIPT (CAT) Test of OP_CAT flag by calling CAT on an empty stack. This does not error because no OP_CAT flag is set so CAT is OP_SUCCESS"
+],
+[
+    [
+        "CAT",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "INVALID_STACK_OPERATION",
+    "TAPSCRIPT Test of OP_CAT flag by calling CAT on an empty stack. This throws an error because OP_CAT flag is set so CAT is executed"
+],
+[
+    [
+        "aa",
+        "bb",
+        "CAT",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT Test of OP_CAT flag by calling CAT on two elements. OP_CAT flag is set so CAT is executed."
+],
+[
+    [
+        "78a11a1260c1101260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT CATs 78a11a1260 and c1101260 together and checks it is EQUAL to stack element 78a11a1260c1101260"
+],
+[
+    [
+        "78a11a1260c1101260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT",
+    "OK",
+    "TAPSCRIPT Test of OP_CAT flag, CATs 78a11a1260 and c1101260 together and checks it is EQUAL to stack element 78a11a1260c1101260. No OP_CAT set so CAT should be OP_SUCCESS."
+],
+[
+    [
+        "",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "EVAL_FALSE",
+    "TAPSCRIPT CATs 78a11a1260 and c1101260 together and checks it is EQUAL to the empty stack element"
+],
+[
+    [
+        "51",
+        "78a11a1260c1101260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUALVERIFY",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT CATS 78a11a1260 and c1101260 together and checks it is EQUALVERIFY to stack element 78a11a1260c1101260"
+],
+[
+    [
+        "51",
+        "c110126078a11a1260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUALVERIFY",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "EQUALVERIFY",
+    "TAPSCRIPT CATs 78a11a1260 and c1101260 together and checks it is EQUALVERIFY to stack element c110126078a11a1260"
+],
+[
+    [
+        "aa",
+        "bb",
+        "CAT 0x4c 0x02 0xaabb EQUAL",
```

Comment:
A concern I have here is that it could (perhaps?) be possible to set a witness element that could both be interpreted as hex and as script. An encoding tag would make that impossible.

Reply:
Thanks for your helpful comments. I addressed this in this PR. Not merged yet, but look it over and see if you have any additional feedback.

Looks like: `"#SCRIPT# CAT CAT DUP DROP 0x4c 0x07 0xeeffeeffaabbff EQUAL"` now

https://github.com/0xBEEFCAF3/bitcoin/pull/7/files#diff-ef0159635ee0fee8864cd9570a04337cabb63e9c49a9e96d7b746194ecaa427d

Reply:
@EthanHeilman Looks great!

Reply:
I implemented the same set of changes for btcd, works well: https://github.com/btcsuite/btcd/pull/2095/commits/4ffc4019ba0e37463649a37a1485dfe294f1675e